voice agent tutorial (#1747)
langchain-ai/docs: src/oss/langchain/voice_agent.mdx

---
title: Build a voice agent with LangChain
sidebarTitle: Voice agent
---

## Overview

Some applications, including customer support, personal assistants, and hands-free interfaces, can benefit from supporting real-time interactions through audio.

> This guide demonstrates how to interact with LangChain [agents](/oss/langchain/agents) through voice channels. It follows the reference architecture implemented in the LangChain repo [here](https://github.com/langchain-ai/voice-sandwich-demo).

Reviewer: ```suggestion
This tutorial demonstrates how to interact with LangChain [agents](/oss/langchain/agents) through voice channels. It follows the reference architecture implemented in the [LangChain repo](https://github.com/langchain-ai/voice-sandwich-demo).
```

> ### Concepts

Reviewer: ```suggestion
## Concepts
```

There are two common architectures for voice agents:

> #### 1. "The Sandwich"

Reviewer: ```suggestion
### 1. "The Sandwich"
```

- Transparent behavior with clear boundaries between components

**Cons:**
- Requires orchestrating multiple services
- Additional complexity in managing the pipeline
- Conversion from speech to text loses information (e.g., tone, emotion)

#### 2. Speech-to-Speech Architecture

Speech-to-speech uses a multimodal model that processes audio input and generates audio output natively.

```mermaid
flowchart LR
    A[User Audio] --> B[Multimodal Model]
    B --> C[Audio Output]
```

**Pros:**
- Simpler architecture with fewer moving parts
- Typically lower latency for simple interactions
- Direct audio processing captures tone and other nuances of speech

**Cons:**
- Limited model options, greater risk of provider lock-in
- Features may lag behind text-modality models
- Less transparency in how audio is processed
- Reduced controllability and customization options

This guide demonstrates the **sandwich architecture** to balance performance, controllability, and access to modern model capabilities. The sandwich can achieve sub-700ms latency with some STT and TTS providers while maintaining control over modular components.

> ### Demo Application Overview

Reviewer: ```suggestion
## Demo application overview
```

We will walk through building a voice-based agent using the sandwich architecture. The agent will manage orders for a sandwich shop. The application will demonstrate all three components of the sandwich architecture, using [AssemblyAI](https://www.assemblyai.com/) for STT and [ElevenLabs](https://elevenlabs.io/) for TTS (although adapters can be built for most providers).

An end-to-end reference application is available in the [voice-sandwich-demo](https://github.com/langchain-ai/voice-sandwich-demo) repository. We will walk through that application here.

The demo uses WebSockets for real-time bidirectional communication between the browser and server. The same architecture can be adapted for other transports like telephony systems (Twilio, Vonage) or WebRTC connections.

## Setup

For detailed installation instructions and setup, see the [repository README](https://github.com/langchain-ai/voice-sandwich-demo#readme).

> ### Architecture Overview

Reviewer: I think it makes more sense to put the Architecture overview under the Demo application overview section and retitle it to "Architecture"

The demo implements a streaming pipeline where each stage processes data asynchronously:

**Client (Browser)**
- Captures microphone audio and encodes it as PCM
- Establishes WebSocket connection to the backend server
- Streams audio chunks to the server in real-time
- Receives and plays back synthesized speech audio

:::python
**Server (Python)**
:::
:::js
**Server (Node.js)**
:::

- Accepts WebSocket connections from clients
- Orchestrates the three-step pipeline:

>   - [STT](#1-speech-to-text): Forwards audio to the STT provider (e.g., AssemblyAI), receives transcript events

Reviewer: ```suggestion
  - [Speech-to-text (STT)](#1-speech-to-text): Forwards audio to the STT provider (e.g., AssemblyAI), receives transcript events
```

  - [Agent](#2-langchain-agent): Processes transcripts with LangChain agent, streams response tokens

>   - [TTS](#3-text-to-speech): Sends agent responses to the TTS provider (e.g., ElevenLabs), receives audio chunks

Reviewer: ```suggestion
  - [Text-to-speech (TTS)](#3-text-to-speech): Sends agent responses to the TTS provider (e.g., ElevenLabs), receives audio chunks
```
